lib/experimental/html/class-wp-html-tag-processor.php

```diff
@@ -1397,6 +1397,24 @@ public function get_attribute( $name ) {
 		}
 
 		$comparable = strtolower( $name );
+
+		if ( 'class' === $name ) {
+			$this->class_name_updates_to_attributes_updates();
+		}
+
+		// If we have an update for this attribute, return the updated value.
+		if ( isset( $this->attribute_updates[ $comparable ] ) ) {
+			$attribute = trim( $this->attribute_updates[ $comparable ]->text );
+			// Is this a boolean attribute?
+			if ( $attribute === $comparable ) {
```

Comment:
this condition isn't obvious to me. in fact I'm struggling to understand how it's working. I know we have this ambiguity where the updates are purely lexical and we've lost the semantics of the type of update by the point we're looking here, and I'm wondering about that.

setting a value to `false` is indistinguishable from removing the same attribute.
removing all remaining classes is the same too as removing the `class` attribute.
setting a value to `true` should produce an empty `text` value I think.

can we make the check more direct? or was the point of the comment to indicate that this is questionable and elicit help in clarifying?

Reply:
A `$attribute_update[ $comparable ]->text` string is of the shape:

1. `my-attr="myvalue"` _(and not just `myvalue`!)_ for non-boolean attributes, and
2. `enabled` for (true) boolean ones.

I wanted a criterion to return `true` for the latter (and to avoid attempting to extract a non-boolean value in that case). Comparing the normalized attribute name (`$comparable`) to `$attribute_update[ $comparable ]->text` (whatever we want to call that thing) seemed like a reliable way to do. (Another option I considered was the absence of an equals sign, but that didn't seem too elegant either.) Given the shape of that thing, I'm not sure if I can think of a more direct check 🤔 -- but happy to do so if you have any ideas. 

> or was the point of the comment to indicate that this is questionable and elicit help in clarifying?

Ah, not really, TBH 😅 FWIW, the [comment a few lines below](https://github.com/WordPress/gutenberg/pull/46680/files#diff-93724b2f6e5dbf16c2b3c0efea57fa442b572f75f325d1cb185cf614b692cf9fR1413) was meant to remind the reader of the shape of `$attribute_update[ $comparable ]->text` (once we've ruled out that it's a boolean one):

```
// $attribute is a 'name="value"' string, so we extract the value.
```

Reply:
I'm worried that this condition is going to collapse as we perform HTML content changes.  we probably need to figure out a much better way of communicating the types of updates before we start stitching and cutting HTML… 🤔

Reply:
Can you maybe give an example of what type of scenario you're concerned about? I can't seem to think of one off the top of my head (which might be due to silent assumptions on my part).

Are we going to allow cutting and stitching _within_ an opening tag? I thought if we limited the interface to methods like `[g|s]et_content_within_balanced_tags`, attribute changes could maybe still use the current format, since they wouldn't be affected -- unless they are for a tag within those balanced tags, but in that case, they'd need to be flushed/wiped anyway, no?

Also, is this something you'd like us to sort out already, or something that can wait until we implement that kind of modifications?

Reply:
at this point the name is bad. it's `lexical_updates` and not `attribute_updates`, so the immediate work you're wanting to do, `set_content_within_balanced_tags` will produce `->text` that doesn't match this pattern.

Reply:
as long as `$comparable` is non-empty then it may be fine as that implies it's an attribute update, vs. `$this->attribute_updates[] = new WP_HTML_Text_Replacement(…)` for content changes

new thoughts:

 - we go with this for now, but note its fragility
 - we add a semantic value to the text replacement which we can query. given the limited and transient nature of these instances I'm not worried about memory use in them.
 - review the whole write-read-write-read cycle as it feels like it's starting to get out of hand. are we avoiding `get_udpated_html()` at the cost of greatly increased complexity in the code.

@adamziel at one point created a `flush()` method (or something similar) which amounted to "flush all enqueued updates" and which we could call whenever. this would make it easier to start all query, update, and seek calls and then never worry about differential accounting as we're doing in this PR

Reply:
Thank you Dennis! I've started [work on `set_content_inside_balanced_tags`](https://github.com/WordPress/gutenberg/pull/47036), and I'm now understanding your thoughts a lot better 😊 

I'm thinking to play a bit more with https://github.com/WordPress/gutenberg/pull/47036 to see if I can massage the updating logic into a certain shape that will be compatible with what I'm trying to do here. I _think_ we have a realistic shot at this; see my notes in that PR's description 😊 If that works out and I can limit the changes transparently to `private` methods, I might try to get those in separately, in order to unblock this PR.

Reply:
```suggestion
			/*
			 * When the lexical update contains just the attribute name,
			 * that attribute value becomes a boolean true. For example:
			 * 
			 * ```php
			 * $p = new WP_HTML_Tag_Processor('<input type="checkbox" />');
			 * $p->set_attribute('checked', true); 
			 * // The lexical update contains just the string "checked"
			 * 
			 * $p->get_attribute('checked');
			 * // returns true thanks to the condition below
			 * 
			 * echo $p->get_updated_html();
			 * // <input type="checkbox" checked />
			 * ```
			 */
			if ( $attribute === $comparable ) {
```

Reply:
@adamziel what did you mean to communicate by including this snippet?